RP-EF read and update related data, concurrency

Rick-Anderson opened the pull request:
[Internal review URL Reading related data](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-rp/read-related-data?branch=pr-en-us-4708)
[Updating related data](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-rp/update-related-data?branch=pr-en-us-4708)
[Concurrency](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/concurrency?branch=pr-en-us-4708)

fixes #4739
Fixes #4740

Rick-Anderson commented:
@ajcvickers hoping you can review this too. You're previous review was fantastic. cc @divega
No hurry. Be sure to use the [Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-rp/read-related-data?branch=pr-en-us-4708)

Rick-Anderson commented:
@ajcvickers hold off for now. I've got customer feedback I'm incorporating. It will probably be ready to review Thursday 9 Nov.

Rick-Anderson commented:
@ajcvickers let me know when you can review this  Parts 6 & 7 

* [Read Related data](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-rp/read-related-data?branch=pr-en-us-4708)
* [Update related data](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-rp/update-related-data?branch=pr-en-us-4708)

cc @divega

Rick-Anderson commented:
@ajcvickers @divega please review
Use the links below to review.
[Internal review URL Reading related data](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-rp/read-related-data?branch=pr-en-us-4708)
[Updating related data](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-rp/update-related-data?branch=pr-en-us-4708)
[Concurrency](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/concurrency?branch=pr-en-us-4708)

Rick-Anderson commented:
@ajcvickers ping

ajcvickers commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -31,57 +32,87 @@ The following illustrations show the completed pages for this tutorial:
 There are several ways that EF can load related data into the navigation properties of an entity:
 
-* Eager loading. When the entity is read, its related data is retrieved. This typically results in a single join query that retrieves all of the data that's needed. Eager loading is specified in EF with the `Include` and `ThenInclude` methods.
+* [Eager loading](https://docs.microsoft.com/ef/core/querying/related-data#eager-loading). Eager loading is when a query for one type of entity also loads related entities. When the entity is read, its related data is retrieved. This typically results in a single join query that retrieves all of the data that's needed. Eager loading is specified with the `Include` and `ThenInclude` methods.
```
EF Core will issue multiple queries for some types of eager loading. This is actually a lot more efficient than was the case for some queries in EF6 where there was a single query.

ajcvickers commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -31,57 +32,87 @@ The following illustrations show the completed pages for this tutorial:
-  zz
-  
-Rather than a single join query, some of the data can be retrieved in separate queries, and EF "fixes up" the navigation properties. "fixes up" means that EF automatically adds the related entities where they belong. Eager loading with separate queries results in multiple queries sent to the DB.
+ The data can be retrieved in separate queries, and EF "fixes up" the navigation properties. "fixes up" means that EF automatically populates the navigation properties. Eager loading with separate queries results in multiple queries sent to the DB.
```
When using separate queries with "Load" this is more like explicit loading than eager loading.

ajcvickers replied:
Although I agree it is somewhat ambiguous. Maybe it should just be its own section?

Rick-Anderson replied:
@ajcvickers  I made a section for this.

ajcvickers commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -31,57 +32,87 @@ The following illustrations show the completed pages for this tutorial:
-* Lazy loading. When the entity is first read, related data isn't retrieved. However, the first time you attempt to access a navigation property, the data required for that navigation property is automatically retrieved. A query is sent to the DB each time you try to get data from a navigation property for the first time. Entity Framework Core 1.0 does not support lazy loading.
+* [Lazy loading](https://docs.microsoft.com/ef/core/querying/related-data#lazy-loading). [EF Core does not currently support lazy loading](https://github.com/aspnet/EntityFrameworkCore/issues/3797). When the entity is first read, related data isn't retrieved. The first time a navigation property is accessed, the data required for that navigation property is automatically retrieved. A query is sen ...
+
+* The `Select` operator loads only the related data needed, and is generally faster than using `Include`. [Loading related data with Select](#select) explains this in more detail.
```
"is generally faster than using Include" @divega @anpete Is this true?

anpete replied:
Not in general. Up until recently projecting navs would often mean n+1 queries, and it still does for sub-queries rooted on collection navs (@maumar is working on it). I guess the case where it could be is for non-collection navs that are shaped down to project only a subset of the related data.

divega replied:
Perhaps the statement came from an EF6-based tutorial?

ajcvickers commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -31,57 +32,87 @@ The following illustrations show the completed pages for this tutorial:
 
-On the other hand, in some scenarios separate queries is more efficient. Eager loading of all related data in one query might cause a very complex join to be generated, which SQL Server can't process efficiently. Or if you need to access an entity's navigation properties only for a subset of a set of the entities you're processing, separate queries might perform better because eager loading of ev ...
+* Eager loading generally offers the best performance.
+* Eager loading sends a single query to the DB. A single query is typically more efficient than separate queries for each entity retrieved.
```
See comment above

anpete replied:
Eager loading sends multiple queries any time a collection nav is being included - one query for the main query and one query of each collection "edge" in the load tree.

ajcvickers commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -31,57 +32,87 @@ The following illustrations show the completed pages for this tutorial:
+For example, suppose that each department has 10 related courses. Eager loading of all related data would result in just a single (join) query and a single round trip to the DB. A separate query for courses for each department would result in 11 round trips to the DB. The extra round trips to the DB are especially expensive when latency is high. Cloud environments can have high DB access latency.
 
-The Course entity includes a navigation property that contains the Department entity of the department that the course is assigned to. To display the name of the assigned department in a list of courses, you need to get the Name property from the Department entity that is in the `Course.Department` navigation property.
+In some scenarios, separate queries is more efficient. Eager loading of all related data might cause a complex join to be generated. SQL Server might not be able to process a complex join efficiently. If only a subset of related data needs to be accessed, separate queries might perform better. Eager loading loads all related data, and the extra data fetching may cost more than separate queries. A ...
```
This sounds like an argument based on EF Classic, rather than EF Core.

Rick-Anderson replied:
@ajcvickers  it doesn't really fit into the article. I'm removing the entire paragraph.

ajcvickers commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -91,93 +122,98 @@ Run the app and select the **Courses** tab to see the list with department names
+[!code-csharp[Main](intro/samples/cu/Pages/Courses/Index.cshtml.cs?name=snippet_RevisedIndexMethod&highlight=4)]
 
-This page reads and displays related data in the following ways:
+The `Select` operator loads only the related data needed, and is generally faster than using `Include`. For single items, like the `Department.Name` it uses a SQL INNER JOIN. For collections it uses another database access, but so does the .`Include` operator on collections.
```
This seems to contradict the single join statements I commented on above. Also, Include may also use an inner join depending on the configuration of the relationships.

ajcvickers commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -199,53 +235,106 @@ You've made the following changes to the existing code:
+
+[!code-csharp[Main](intro/samples/cu/Pages/Instructors/IndexSingle.cshtml.cs?name=snippet_single&highlight=21,28-29)]
+
+The preceding `Single` approach provides no benefits over using `Where`. Some developers prefer the `Single` approach style.
```
This is causing the whole list to be loading into memory and then the Single call is executing in-memory (with LINQ to Objects) against that list. Is there are reason to do this rather than using SingleAsync on the query itself so that the selection can happen in the database?

Rick-Anderson replied:
@ajcvickers  see [code here](https://github.com/aspnet/Docs/blob/e0f83039d06bfec8662656e92c1355f5c31d3d03/aspnetcore/data/ef-rp/intro/samples/cu/Pages/Instructors/IndexSingle.cshtml.cs)

Yes, need to load entire list. This is the index page for instructors. The other sub-queries are if you have selected an instructor on the page.

divega commented on aspnetcore/data/ef-rp/concurrency.md:
```diff
@@ -0,0 +1,333 @@
+If concurrency detection is not enabled, when concurrent updates occur:
+
+* The last update wins. That is, the last update values are saved to the DB.
+* The first of the current updates are lost. That's generally not what you want.
```
> That's generally not what you want.

I currently disagree with this statement. I think we must recognize that concurrency is useful only in a subset of scenarios. When it matters, you can really use help from your persistence framework. I.e. concurrency control features can help make applications that handle multiple users almost as easy to code as single-user logic (and that is why we support it and why it is important to document how it works), but still we don't do any concurrency control by default.

Whether you care about concurrency is going to depend on whether there are rules that constrain what changes to the database are valid, e.g. if an application performs a debit transaction against a bank account, it matter that things balanced out at the end and that there is a non-negative balance at the end. If you validate that this is the case in the unit-of-work, then you can use concurrency to make sure no other debit transaction will be able to happen concurrently on the same account. But within the same application, you may not care if the description of the account is updated. @

divega commented on aspnetcore/data/ef-rp/concurrency.md:
```diff
@@ -0,0 +1,333 @@
+DB locking, also called pessimistic concurrency, is a concurrency approach. For example, before a row is read from the db, the app can request a lock for read-only or for update access. If a row is locked for update access:
+
+* Other users can't lock the row for read-only access. That is, other users can't read the data.
+* Other users can't lock the row for update access.
```
This seems a bit confusing to me :smile: Is this definition of what "locked for update access" means coming from somewhere else? E.g. update locks in SQL Server are compatible with shared locks, which is what I believe you are referring to with "locked for read-only acesss".

divega commented on aspnetcore/data/ef-rp/concurrency.md:
```diff
@@ -0,0 +1,333 @@
+* Other users can't lock the row for read-only access. That is, other users can't read the data.
+* Other users can't lock the row for update access.
+
+If a row is locked for read-only access:
```
Is this shared locks?

divega commented on aspnetcore/data/ef-rp/concurrency.md:
```diff
@@ -0,0 +1,333 @@
+
+### Detecting concurrency conflicts on a row
+
+To detect concurrency conflicts, a [rowversion](https://docs.microsoft.com/sql/t-sql/data-types/rowversion-transact-sql) tracking column is added to the model. `rowversion` is used to determine that an entity has not been changed since it was fetched from the DB. The DB generates a sequential `rowversion` number that's incremented each time the row is updated. In an `Update` or `Delete` command,  ...
```
Is this document intended to be only about how to achieve things in SQL Server? If no, we should clarify here that we are talking about a SQL Server feature and that other databases may not provide a similar feature.

Rick-Anderson commented:
@divega  address all your comments. Removed entire section on locking given the tutorial doesn't use locking.  Removed all the incorrect faster statements.

>refers to EF Core as just "EF" a lot. 

Ouch. I must have fixed 50 occurrences in the series. 

>This has a lot of overlap with conceptual topics of EF Core. There is a trade off between having all this information here and the risk of it being wrong or becoming outdated. It also makes it very long. Should we tune it down to be shorter and less deep?

This series is a copy of the EF/MVC series. I changed from MVC to Razor Pages, shortened each section and removed the last 2 tutorials. The original series had a good review from your team (MVC3/EF4) and has since been updated without much review (although we did ask for reviews :))

The series is extremely popular with customers - currently getting 125K+ views/month. The MVC5/EF6 series gets slightly better views. Once more customers migrate to .NET Core we expect readership to double.

I've never seen any tutorial MS produced get reviews as good as this series. 

I'd like to:

* Keep  this series for now
* Work on trimming it down (and pointing to the EF docs where I trim it down).
* Retire the [ASP.NET Core 2 MVC/EF Core 2](https://docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/intro) series 
* Consider retiring this series with .NET Core 3

I've got alot of bugs to fix in the EF Core/MVC Core version from feedback on this version.

Rick-Anderson commented:
@divega Let me know if my changes work. I'm happy to trade your team 3 hours of work for every hour you put in this series.

ajcvickers commented:
@Rick-Anderson Diego's comments aside, this looks good to me.

divega commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -31,34 +32,41 @@ The following illustrations show the completed pages for this tutorial:
 
 There are several ways that EF can load related data into the navigation properties of an entity:
```
EF Core

divega commented on aspnetcore/data/ef-rp/read-related-data.md:
```diff
@@ -31,34 +32,41 @@ The following illustrations show the completed pages for this tutorial:
   ![Explicit loading example](read-related-data/_static/explicit-loading.png)
 
-* Lazy loading. When the entity is first read, related data isn't retrieved. However, the first time you attempt to access a navigation property, the data required for that navigation property is automatically retrieved. A query is sent to the DB each time you try to get data from a navigation property for the first time. Entity Framework Core 1.0 does not support lazy loading.
+* Lazy loading. [EF Core does not currently support lazy loading](https://github.com/aspnet/EntityFrameworkCore/issues/3797). When the entity is first read, related data isn't retrieved. The first time a navigation property is accessed, the data required for that navigation property is automatically retrieved. A query is sent to the DB each time a navigation property is accessed for the first tim ...
```
This was supposed a list of ways EF Core **can** load related data. Moves this whole paragraph to an [!Information] or [!Note] thing?

divega replied:
deleting comments I made while looking at just this commit. Sorry.

divega commented:
> although we did ask for reviews :)

If I am not mistaken, I did participate in a review of this before, once or maybe even twice. It is only every time we review it we find new stuff :smile_cat:, but I guess some things are new too. 

Anyway, thanks for your patience @Rick-Anderson.
